lib/TWCManager/TWCMaster.py

```diff
@@ -244,11 +243,49 @@ def getScheduledAmpsMax(self):
             return 0
 
     def getScheduledAmpsStartHour(self):
-        return self.settings.get("scheduledAmpsStartHour", -1)
+        return int(self.settings.get("scheduledAmpsStartHour", -1))
+
+    def getScheduledAmpsTimeFlex(self):
+        startHour = self.getScheduledAmpsStartHour()
+        days = self.getScheduledAmpsDaysBitmap()
+        if (
+            startHour >= 0
+            and self.getScheduledAmpsFlexStart()
+            and self.countSlaveTWC() == 1
+        ):
+            # Try to charge at the end of the scheduled time
+            slave = next(iter(self.slaveTWCs.values()))
+            vehicle = slave.getLastVehicle()
+            if vehicle != None:
+                amps = self.getScheduledAmpsMax()
+                watts = self.convertAmpsToWatts(amps) * self.getRealPowerFactor(amps)
+                hoursForFullCharge = self.getScheduledAmpsBatterySize() / (watts / 1000)
+                realChargeFactor = (vehicle.chargeLimit - vehicle.batteryLevel) / 100
```

Comment:
It's slightly more complicated than this, because the car's _current_ charge limit might not be the charge limit applied by the Scheduled Charging policy.  You should check whether the Scheduled Charging policy has a limit set, and if not, use the car's current limit.

Even that's not totally correct, since a custom policy could check whether it's within the Scheduled window as one of its criteria; I haven't done that, but one could, especially if they're doing a fully custom policy set.  You'd almost want either the limit or the policy object to be an argument to this function, so it could always reference the right thing, but then you need to figure out which policy you're talking about in places where it's called.

Reply:
Yes, I know - it is more complicated. But that's why I simplified it - If you us scheduled charging - you have to know you charging speed and it works for exactly one car. I don't want to start charge logic of the TWCManager. Added some documentation about this.

The second one is only a problem if you use checkScheduledCharging in your policy and setting flexStart - so in my opinion if you want to Schedule Charge and want to have a flex start - it works as it should.

Reply:
I don't think that mitigates the point about the charge limit, though.  I leave my car's limit set to 100% when it's away from home, and rely on TWCManager to set it to what the policy is going to use (e.g. Scheduled Charging to 50%).  By looking at the car's current limit rather than the Scheduled Charging policy's limit, you're going to get into this loop:

- Car's limit is 100%; calculate the start time needed to reach 100%
- Scheduled Charging begins and sets the limit to 50%
- The time needed to reach 50% means not starting yet, so Scheduled Charging is no longer the selected policy.
- Non-Scheduled Charging begins and removes the limit override; the limit returns to 100%
- Scheduled Charging begins, because we're past the start time needed to reach 100%
- ...

I've gone ahead and merged this at @ngardiner's suggestion, so I'll file a follow-up issue for that.

Reply:
Now I get it, sorry. So a change could be that it starts the scheduled charging always at the begining. But stops then when it gets the flex time back.